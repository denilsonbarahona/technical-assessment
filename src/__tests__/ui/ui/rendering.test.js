import React from "react";
import Header from "../../../components/header";
import News from "../../../containers/news";
import { render, waitFor } from "@testing-library/react";
import { screen } from "@testing-library/dom";
import ProviderMock from "../../../__mocks__/providerMock";
import MockResponse from "../../../__mocks__/fetchResponseMock";

function MockingFetching(status) {
  const mockEntries = [{ isIntersecting: false }];
  window.IntersectionObserver = jest.fn(() => ({
    observe: jest.fn(),
    unobserve: jest.fn(),
  }));
  global.fetch = jest.fn(() =>
    Promise.resolve({
      ok: status,
      json: () => Promise.resolve(MockResponse),
    })
  );
  jest.spyOn(document, "querySelector").mockReturnValueOnce(mockEntries);
}

describe("check rendering of elements", () => {
  test("check rendering of header", () => {
    render(<Header />);
    const header = screen.queryAllByRole("banner");
    expect(header).toHaveLength(1);
  });
});

describe("check rendering of loader and dropdown", () => {
  MockingFetching(true);
  test("check appearance and disappearance of loading", async () => {
    const { container, getByRole } = render(
      <ProviderMock>
        <News />
      </ProviderMock>
    );
    const loader = container.getElementsByClassName("loader");
    expect(loader.length).toBe(1);
    await waitFor(() => {
      expect(loader.length).toBe(0);
      expect(getByRole("listbox")).toBeInTheDocument();
    });
  });
});

describe("rendering News", () => {
  test("checking appearance of news item when fetching status 200", async () => {
    MockingFetching(true);
    const { container } = render(
      <ProviderMock>
        <News />
      </ProviderMock>
    );
    const newsItems = container.getElementsByClassName("news-item");
    await waitFor(() => {
      expect(newsItems.length).toBeGreaterThan(0);
    });
  });

  test("checking appearance of notification error when fetching fails", async () => {
    MockingFetching(false);
    const { getByRole } = render(
      <ProviderMock>
        <News />
      </ProviderMock>
    );

    await waitFor(() => {
      expect(getByRole("alert")).toBeInTheDocument();
    });
  });
});
