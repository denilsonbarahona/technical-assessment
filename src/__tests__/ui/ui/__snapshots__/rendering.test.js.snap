// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering News testing snapshot of header 1`] = `
<div>
  <header
    class="header"
  >
    <h1
      class="header__title"
    >
      Hacker News
    </h1>
  </header>
</div>
`;
